pkg/api/kptfile/v1alpha2/types.go

```diff
@@ -0,0 +1,201 @@
+	// merges the changes into the local package.
+	ResourceMerge UpdateStrategyType = "resource-merge"
+	// FastForward fails without updating if the local package was modified
+	// since it was fetched.
+	FastForward UpdateStrategyType = "fast-forward"
+	// ForceDeleteReplace wipes all local changes to the package.
+	ForceDeleteReplace UpdateStrategyType = "force-delete-replace"
+)
+
+// Upstream is a reference to where the package is fetched from.
+type Upstream struct {
+	// Type is the type of origin.
+	Type OriginType `yaml:"type,omitempty"`
+
+	// Git contains information on the origin of packages fetched from a git repository.
+	Git Git `yaml:"git,omitempty"`
+
+	// UpdateStrategy defines how a package is updated from upstream.
+	UpdateStrategy UpdateStrategyType `yaml:"updateStrategy,omitempty"`
+}
+
+// PackageMeta contains metadata such as license, documentation, etc.
+// These fields are not used for any functionality in kpt and are simply passed through.
+type PackageMeta struct {
+	// URL is the location of the package.  e.g. https://github.com/example/com
+	URL string `yaml:"url,omitempty"`
+
+	// Email is the email of the package maintainer
+	Email string `yaml:"email,omitempty"`
+
+	// License is the package license
+	License string `yaml:"license,omitempty"`
+
+	// Version is a logical package version (ignored by kpt)
+	Version string `yaml:"version,omitempty"`
+
+	// Tags enables humans and tools to attach arbitrary package metadata.
+	Tags []string `yaml:"tags,omitempty"`
+
+	// Man is the path to documentation about the package
+	Man string `yaml:"man,omitempty"`
+
+	// ShortDescription contains a short description of the package.
+	ShortDescription string `yaml:"shortDescription,omitempty"`
+}
+
+// Subpackages declares a local or remote subpackage.
+type Subpackage struct {
+	// Name of the immediate subdirectory relative to this Kptfile where the suppackage
+	// either exists (local subpackages) or will be fetched to (remote subpckages).
+	// This must be unique across all subpckages of a package.
+	LocalDir string `yaml:"localDir,omitempty"`
+	// Whether a subpackage is local or remote is determined by whether Upstream is specified.
+	// Upstream is a reference to where the package is fetched from.
+	Upstream Upstream `yaml:"upstream,omitempty"`
+}
+
+// Git contains information on the origin of packages fetched from a git repository.
+type Git struct {
+	// Commit is the git commit that the package was fetched at
+	Commit string `yaml:"commit,omitempty"`
+
+	// Repo is the git repository the package was cloned from.  e.g. https://
+	Repo string `yaml:"repo,omitempty"`
+
+	// Directory is the sub directory of the git repository that the package was cloned from
+	Directory string `yaml:"directory,omitempty"`
+
+	// Ref is the git ref the package was cloned from
+	Ref string `yaml:"ref,omitempty"`
+}
+
+// Pipeline declares a pipeline of functions used to mutate or validate resources.
+type Pipeline struct {
```

Comment:
Can we use the `Pipeline` and `Function` defined in `internal/pipeline`? Otherwise we need to move a lot of codes here.

Reply:
We are looking to move away from unmarshalling the Kptfile into structs, so I don't think we need to worry about it. I think this structure is mostly about communicating the structure.

Reply:
The schema should be used in code. Discussion of how it's marshalled/unmarshalled from YAML is orthogonal to this. In addition, Kptfile needs to eventually have a CRD which can be generated from this.

@Shell32-Natsu This is mostly the same as the struct in internal/pipeline, so refactoring the pipeline code to use this should follow easily.

Reply:
I think @Shell32-Natsu point is about code organization, are we moving all the types definition (atleast referred by Kptfile) in one place under `v1alpha2` or just having root type here that can refer to types defined in other packages. There is some code that is written in OOP style where methods are implemented on types for ex. `Function` and `Pipeline`, so it impacts that.

Having types in one place here might be simpler for the main types that Kptfile directly uses.  @Shell32-Natsu pipeline pkg's code will use the types defined here.

Reply:
Yes, we want to have all the types in Kptfile be self-contained in this package. The question is really about method receivers on these types. 

This is actually a much-asked feature require in Go as other language provide type extension:

https://github.com/golang/go/issues/21401

Choices are:

1. Use functions instead of method receivers
2. Put method receivers in this package
3. Type alias or embed these types. See https://stackoverflow.com/a/28800807

If the method receivers are generic and reusable across many features, 2 seems reasonable. If on the other hand, they're specific to one feature (e.g. Pipeline validation), then 1 and 3 are reasonable.

Reply:
closing the loop here: @Shell32-Natsu and me had an offline chat about this. There are very few methods defined on Pipeline and Function and turning them in to functions won't be a big deal. So all set.